pysph/cpy/jit.py

```diff
@@ -0,0 +1,523 @@
+    def __init__(self, reduce_expr, map_func=None, dtype_out=np.float64,
+                 neutral='0', backend='cython'):
+        backend = array.get_backend(backend)
+        self.tp = Transpiler(backend=backend)
+        self.backend = backend
+        self.func = map_func
+        if map_func is not None:
+            self.name = 'reduce_' + map_func.__name__
+        else:
+            self.name = 'reduce'
+        self.reduce_expr = reduce_expr
+        self.dtype_out = dtype_out
+        self.type = dtype_to_ctype(dtype_out)
+        if backend == 'cython':
+            # On Windows, INFINITY is not defined so we use INFTY which we
+            # internally define.
+            self.neutral = neutral.replace('INFINITY', 'INFTY')
+        else:
+            self.neutral = neutral
+        self._config = get_config()
+        self.cython_gen = CythonGenerator()
+        self.queue = None
+
+    def get_type_info_from_args(self, *args):
+        type_info = {}
+        arg_names = getargspec(self.func)
+        if 'i' in arg_names:
+            arg_names.remove('i')
+            type_info['i'] = 'int'
+        for arg, name in zip(args, arg_names):
+            arg_type = get_ctype_from_arg(arg)
+            if not arg_type:
+                arg_type = 'double'
+            type_info[name] = arg_type
+        return type_info
+
+    @memoize
+    def _generate_kernel(self, *args):
+        if self.func is not None:
+            arg_types = self.get_type_info_from_args(*args)
+            helper = AnnotationHelper(self.func, arg_types)
+            helper.annotate()
+            self.func = helper.func
+            external_funcs = {}
+            gather_external_funcs(helper, external_funcs)
+            self.tp.external_funcs = external_funcs
+        return self._generate()
+
+    def _massage_arg(self, x):
+        if isinstance(x, array.Array):
+            return x.dev
+        elif isinstance(x, np.ndarray):
+            return x
+        else:
+            return np.asarray(x)
+
+    def __call__(self, *args, **kw):
+        c_func = self._generate_kernel(*args)
+        c_args = [self._massage_arg(x) for x in args]
+
+        if self.backend == 'cython':
+            size = len(c_args[0])
+            c_args.insert(0, size)
+            return c_func(*c_args, **kw)
+        elif self.backend == 'opencl':
+            result = c_func(*c_args, **kw)
+            self.queue.finish()
+            return result.get()
+        elif self.backend == 'cuda':
+            import pycuda.driver as drv
+            event = drv.Event()
+            result = c_func(*c_args, **kw)
+            event.record()
+            event.synchronize()
+            return result.get()
+
+
+class ScanJIT(parallel.ScanBase):
+    def __init__(self, input=None, output=None, scan_expr="a+b",
+                 is_segment=None, dtype=np.float64, neutral='0',
+                 complex_map=False, backend='opencl'):
+        backend = array.get_backend(backend)
```

Comment:
This is rather similar to the ReductionJIT so should one be a subclass?

Reply:
Note, I know that a lot of the earlier code I wrote requires a massive refactor as that was my first cut... :)